types/parcel-bundler/index.d.ts

```diff
@@ -0,0 +1,167 @@
+     *
+     * @default false
+     */
+    contentHash?: boolean;
+    /**
+     * Minify files
+     *
+     * @default process.env.NODE_ENV === 'production'
+     */
+    minify?: boolean;
+    /**
+     * Turn on experimental scope hoisting/tree shaking flag, for smaller production bundles
+     *
+     * @default false
+     */
+    scopeHoist?: boolean;
+    /**
+     * @default "browser"
+     */
+    target?: "browser" | "node" | "electron";
+    /**
+     * Define a custom {key, cert} pair
+     *
+     * Use true to generate one or false to use http
+     */
+    https?:
+        | true
+        | false
+        | {
+              /**
+               * Path to custom certificate
+               *
+               * @default "./ssl/c.crt"
+               */
+              cert?: string;
+              /**
+               * Path to custom key
+               *
+               * @default "./ssl/k.key"
+               */
+              key?: string;
+          };
+    /**
+     * 3 = log everything, 2 = log warnings & errors, 1 = log errors
+     *
+     * @default 3
+     */
+    logLevel?: 3 | 2 | 1;
+    /**
+     * The port the HMR socket runs on
+     *
+     * Defaults to a random free port (0 in node.js resolves to a random free port)
+     *
+     * @default 0
+     */
+    hmrPort?: 0 | number;
+    /**
+     * Enable or disable sourcemaps
+     *
+     * Defaults to enabled (not supported in minified builds yet)
+     *
+     * @default true
+     */
+    sourceMaps?: boolean;
+    /**
+     * A hostname for hot module reload
+     *
+     * @default ""
+     */
+    hmrHostname?: string;
+    /**
+     * Prints a detailed report of the bundles, assets, filesizes and times
+     *
+     * Reports are only printed if watch is disabled
+     *
+     * @default false
+     */
+    detailedReport?: boolean;
+}
+
+export type ParcelAsset = any;
+
+export interface ParcelBundle {
+    /**
+     * The type of assets it contains (e.g. js, css, map, ...)
+     */
+    type: string;
+    /**
+     * The name of the bundle (generated using Asset.generateBundleName() of entryAsset)
+     */
+    name: string;
+    /**
+     * The parent bundle, is null in case of the entry bundleany
+     */
+    parentBundle?: any;
+    /**
+     * The entryPoint of the bundle, used for generating the name and gathering assets.
+     */
+    entryAsset: any;
+    /**
+     * A Set of all assets inside the bundle
+     */
+    assets: Set<any>;
+    /**
+     * A Set of all sibling bundles
+     */
+    siblingBundles: Set<any>;
+    /**
+     * A Map<String(Type: js, css, map, ...), Bundle> of all sibling bundles
+     */
+    siblingBundlesMap: Map<string, ParcelBundle>;
+    /**
+     * A Map<Asset, number(line number inside the bundle)> of all the locations of the assets inside the bundle, used to generate accurate source maps
+     */
+    offsets: Map<ParcelAsset, number>;
+}
+
+export default class ParcelBundler {
```

Comment:
This [should be](https://github.com/parcel-bundler/parcel/blob/master/src/Bundler.js#L788)
```ts
class ParcelBundler {
    constructor(entryFiles?: string | string[], options?: ParcelOptions);
    bundle(): Promise<ParcelBundle>;
}
exports = ParcelBundler;
```
and if your runtime module loader makes a cjs module available as a default, use the `esModuleInterop` and `allowSyntheticDefaultImports` compiler options.

Reply:
Hi,

I tried and i got this errors when running `npm run test`

```
Error: ./types/parcel-bundler/index.d.ts:163:1
ERROR: 163:1  expect                           TypeScript@next compile error:
A 'declare' modifier is required for a top level declaration in a .d.ts file.
ERROR: 163:7  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 169:1  expect                           TypeScript@next compile error:
Statements are not allowed in ambient contexts.
ERROR: 169:1  expect                           TypeScript@next compile error:
Cannot find name 'exports'.
ERROR: 169:1  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect                           TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 9:23   expect                           TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

So, i tried to add a `declare`

```diff
-class ParcelBundler {
+declare class ParcelBundler {
     constructor(entryFiles?: string | string[], options?: ParcelOptions);
 
     bundle(): Promise<ParcelBundle>;
 }
 
 exports = ParcelBundler;
```
And i got this errors

```
Error in parcel-bundler
Error: ./types/parcel-bundler/index.d.ts:163:1
ERROR: 163:1  strict-export-declare-modifiers  Prefer 'export' to 'declare' in an external module. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 169:1  expect                           TypeScript@next compile error:
Statements are not allowed in ambient contexts.
ERROR: 169:1  expect                           TypeScript@next compile error:
Cannot find name 'exports'.
ERROR: 169:1  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect                           TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 9:23   expect                           TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

I don't understand what i did wrong, i'm new in the TypeScript ecosystem, so i don't understand everything

What should i do ?

Reply:
Ah, my fault. It's `export =` not `exports =`. I typo'd.

Reply:
i tried this

```typescript
declare class ParcelBundler {
    constructor(entryFiles?: string | string[], options?: ParcelOptions);

    bundle(): Promise<ParcelBundle>;
}

export = ParcelBundler;
```

and got this

```
Error in parcel-bundler
Error: ./types/parcel-bundler/index.d.ts:169:1
ERROR: 169:1  expect  TypeScript@next compile error:
An export assignment cannot be used in a module with other exported elements.

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 1:25   expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no exported member 'ParcelOptions'.
ERROR: 1:46   expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' resolves to a non-module entity and cannot be imported using thisconstruct.
ERROR: 9:23   expect  TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

Reply:
Righto, you'll need to wrap the other elements in a `declare namespace ParcelBundler` declaration so they all become elements under the export assigned class/namespace merge.

Reply:
i tried and this didn't work, you can see my last commits